### DISCUSSION TITLE
Using Azure OpenAI and External OpenAI in the UI

### DISCUSSION THREAD

**Discussion by sheldonhull** [open-webui/open-webui#849] [Q&A] [answered]
I'm slightly confused. The external tab has the option for OpenAI with the endpoint and key. I've configured this thinking maybe I could toggle between ollama, and also using a private Azure OpenAI endpoint, but I don't know if there's any way to choose the Azure OpenAI model in my chat itself. Instead it's only Ollama models listed. 

If I'm missing something, or that feature is for something different, feel free to point me to the docs so I can better figure it out. 😁
![CleanShot 2024-02-22 at 15 34 18@2x](https://github.com/open-webui/open-webui/assets/3526320/ff42a2a8-2dab-41ef-b1f4-f0c615d9b6e8)

Thanks!

**Comment by justinh-rahb** [comment, marked as answer]
This is still an ongoing issue:
- #437

**Reply by sheldonhull** [reply]
so to confirm that UI option is unused right now and it's just displayed because it's not hidden behind any type of feature flag? 
pretty sure the URL path I have is fine except I might need to edit the local code to append the version of the API. I just wasn't sure if that external usage was functional yet.

**Reply by justinh-rahb** [reply]
It works with OpenAI-compatible APIs that don't require the version such as Mistral or LiteLLM for example, just not Azure right now. PRs welcome!

**Reply by carhensi** [reply]
I've just opened a discussion [here](https://github.com/open-webui/docs/discussions/530). I'm awaiting feedback to hopefully create a PR after a positive discussion.

**Comment by uright** [comment]
@sheldonhull here's a [blog post](https://uright.ca/blogs/2024-07-29-unlocking-the-power-of-azure-openai-on-open-webui/) that provide some simple instructions to get Azure OpenAI working with LiteLLM.

**Comment by dhuesca** [comment]
any updates on this? 🤔
At the company they gave us powerful laptops so I run some models locally, but also have Azure OpenAI service, I would like to switch between local and cloud based solutions

**Comment by jooneyp** [comment]
I can't believe this feature is still not supported and cannot use my AzureOpenAI credit here..

**Reply by coolspot18** [reply]
Same... was hoping to leverage my credits!

**Comment by vlreinier** [comment]
This is actually supported by using a function. Create a function to connect models from different vendors. Example function for Azure OpenAI: https://openwebui.com/f/nomppy/azure

**Reply by mittalpatel130** [reply]
@vlreinier - How and where do we call this function? Do we have to deploy this as a proxy and also write additional functions for chat completions, etc.?

**Reply by jooneyp** [reply]
> This is actually supported by using a function. Create a function to connect models from different vendors. Example function for Azure OpenAI: https://openwebui.com/f/nomppy/azure

Thanks! This solution works like a charm :)

**Reply by Damsjov** [reply]
Great solution. However I must be doing something wrong - upon uploading this as a function it is not detected by Open WebUI. Is there anything that needs to be done first?

Thanks!

**Reply by amrothemich** [reply]
Same question!

**Reply by jan-koch** [reply]
same here, how do I use that function?

**Reply by CellLz** [reply]
> This is actually supported by using a function. Create a function to connect models from different vendors. Example function for Azure OpenAI: https://openwebui.com/f/nomppy/azure

Although this function can now use Azure's OpenAI API, it currently only supports the traditional format of the API. It requires the apikey, endpoint, development name, and api version to be input separately, and then automatically concatenates them into a complete URL. During this process, adding or missing a '/' in the endpoint will cause an error. Additionally, for Deepseek's model, its URL address is different from the OpenAI API format, and the final URL address can be directly obtained from the Azure control panel, so it is neither necessary nor possible to input it separately. Therefore, could it be modified to a code that only requires inputting the URL and apikey to be used? Thank you!

**Comment by ruudvanstrijp** [comment]
Here is a good guide how to do it as well: https://techcommunity.microsoft.com/blog/educatordeveloperblog/step-by-step-integrate-ollama-web-ui-to-use-azure-open-ai-api-with-litellm-proxy/4386612

**Reply by unicorn667** [reply]
Here is another example 
https://uright.ca/blogs/2024-07-29-unlocking-the-power-of-azure-openai-on-open-webui/

I was searching a solution to use Stable-Diffusion-3.5-Large from Azure.

**Reply by a523** [reply]
> Here is a good guide how to do it as well: https://techcommunity.microsoft.com/blog/educatordeveloperblog/step-by-step-integrate-ollama-web-ui-to-use-azure-open-ai-api-with-litellm-proxy/4386612

I have tried this ,  as the used litellm, this is very slowly.

**Reply by CellLz** [reply]
This third-party function of openwebui is the ultimate solution for integrating Azure OpenAI API. It can adapt to all formats of access addresses and keys, and automatically supports the display and output of chain of thinking.
https://openwebui.com/f/owndev/azure_ai

**Comment by arkodeep3404** [comment]
is azure openai available yet??

**Reply by carhensi** [reply]
Like mentioned in the answer above, I've just [opened a discussion](https://github.com/open-webui/docs/discussions/530) for my hopefully upcoming PR.
